rfcs/constrained-generics.md

```diff
@@ -0,0 +1,49 @@
+# Constrained Generics
+
+## Summary
+
+Allows you to give a function generic a constrained type.
+
+## Motivation
+
+Right now, when you create a generic function, the generic paramter is assumed to be everything. This creates problems, especially with tables
+```lua
+local function SomethingWithAGenericTable<T>(tab: T)
+  for _, v in t do --Cannot call non-function T (T is not a table)
+
+  end
+end
+```
+
+This constraining is mainly for concrete table types and functions, since types like `number`, `string`, and `boolean` are literal types.
+
+## Design
+
+The way to implement this is to have new syntax under the generic's brackets: `<T: typedef>`. This would constrain the type from the generic def and give the typechecker better hints.
+```lua
+local function SomethingWithAGenericTable<T: {[any]: any}>(tab: T)
+  for _, v in t do
+    print(v)
+  end
+end
+
+SomethingWithAGenericTable({meow = 5}) --OK
+SomethingWithAGenericTable(5) --not OK
+```
+
+> TODO: Partially constrained generics, where part of the type is known but can be expaned upon, similar to TypeScript's `extends` keyword.
```

Comment:
What do you have in mind with this exactly? Bounding the type to be a table that has at least some properties? If so, that's already the current meaning with width subtyping in place (though my proposal for exact table types would put that under different syntax). Is there another partial-constraint you had in mind though?

Reply:
I misunderstood the extends keyword from ts lol.

So, lets assume the following, you have an arbitrary dictionary where you know that is has the key "length"

With the proposed syntax, we can either constrain this with {[string]: any}, or {length: number, [string]: any}

This would help the autocomplete since now we know that the generic has at least "length" on the dictionary. It could also be used with typechecking

```lua
consumeLength({}) --not OK
consumeLength({length = 5}) --OK
consumeLength({length = 3, meow = true}) -- OK
```

Reply:
A function that takes an argument of the type `{length: number}` will currently accept any table with at least the field "length." This was proposed and implemented in #38 as the behavior for table types. So, you don't need constrained generics for that at all, and if we added constrained generics, we wouldn't need a special keyword or anything since it's already the meaning of the type. i.e. if you had a constrained function like `for<T>(T) -> T where T: {length: number}` that would tell us that the table type has to be exactly the same in the return type compared to the parameter type and further that it _must_ have a field `length` of type `number`, though it says nothing else about the table type. This would be more obvious too with the proposal I made in #624 where you'd instead write `{ length: number, ...}`for inexact table types.